bn_mp_set_double.c

```diff
@@ -41,19 +54,119 @@ int mp_set_double(mp_int *a, double b)
       return res;
    }
 
-   if (((cast.bits >> 63) != 0ULL) && !IS_ZERO(a)) {
-      a->sign = MP_NEG;
+   if (((cast.bits >> PAYLOAD) != 0ULL) && (mp_iszero(a) == MP_NO)) {
+      SIGN(a) = MP_NEG;
    }
 
    return MP_OKAY;
 }
+
+#else
+static double s_math_h_less_frexp(double x, int *exp)
+{
+   int exponent = 0, i;
+   double high = 2.0;
+   double low = 0.5;
+
+   if (x >= 1.0) {
+      for (i = 0; x >= high; i++) {
+         exponent += (1 << i);
+         x *= low;
+         low *= low;
+         high *= high;
+      }
+      if (x < 0.5) {
+        while (x < 0.5) {
+          x *= 2.0;
+          exponent--;
+        }
+      }
+      else {
+         while (x >= 1.0) {
+            x /= 2.0;
+            exponent++;
+         }
+      }
+   }
+   *exp = exponent;
+   return x;
+}
+
+int mp_set_double(mp_int *a, double b)
+{
+   int exp = 0, res = MP_OKAY, sign = MP_ZPOS;
+
+   /* Check for NaN */
+   /* TODO: there might be some ompilers who do not return true in case of NaN */
+   if (b != b) {
+      return MP_VAL;
+   }
+
+   if (b < 0.0) {
+      b = b * (-1.0);
+      sign = MP_NEG;
+   }
+
+   /* Check for infinity */
+   if ( b > DBL_MAX) {
+      return MP_VAL;
+   }
+
+   mp_zero(a);
+   /* Numbers smaller than 1 truncate to zero */
+   if (b < 1.0) {puts("b < 1.0");
```

Comment:
PUTS, please grep for such things

Reply:
I do, normally, so it can be that I commited the wrong version here, will check. (had a bit of trouble with merging with the HEAD of develop, might be a residue of that fight)